Assets/MRTK/Providers/LeapMotion/LeapMotionArticulatedHand.cs

```diff
@@ -0,0 +1,299 @@
+        public override MixedRealityInteractionMapping[] DefaultInteractions => handDefinition?.DefaultInteractions;
+
+#if LEAPMOTIONCORE_PRESENT
+
+        protected LeapMotionDeviceManager LeapMotionDeviceManager => CoreServices.GetInputSystemDataProvider<LeapMotionDeviceManager>();
+
+        protected AttachmentHands AttachmentHands => LeapMotionDeviceManager.attachmentHands;
+
+        // Attachment hand needed for this hand, left or right hand
+        private AttachmentHand attachmentHand = null;
+
+        // Number of joints in an MRTK Tracked Hand
+        protected static readonly int jointCount = Enum.GetNames(typeof(TrackedHandJoint)).Length;
+
+        // Joint poses of the MRTK hand based on the leap hand data
+        protected readonly Dictionary<TrackedHandJoint, MixedRealityPose> jointPoses = new Dictionary<TrackedHandJoint, MixedRealityPose>();
+
+        private TrackedHandJoint[] metacarpals = new TrackedHandJoint[]
+        {
+            TrackedHandJoint.ThumbMetacarpalJoint,
+            TrackedHandJoint.IndexMetacarpal,
+            TrackedHandJoint.MiddleMetacarpal,
+            TrackedHandJoint.RingMetacarpal,
+            TrackedHandJoint.PinkyMetacarpal
+        };
+
+        /// <summary>
+        /// If true, the current joint pose supports far interaction via the default controller ray.  
+        /// </summary>
+        public override bool IsInPointingPose => handDefinition.IsInPointingPose;
+
+        /// <summary>
+        /// If true, the hand is in air tap gesture, also called the pinch gesture
+        /// </summary>
+        public bool IsPinching => handDefinition.IsPinching;
+
+        private void SetAttachmentHands()
+        {
+            // Only get the attachment hands if the application is playing. The attachment hands are added at runtime and do not exist in edit mode.
+            if (Application.isPlaying)
+            {
+                // Get the attachment hand data based on controller handedness
+                attachmentHand = Array.Find(AttachmentHands.attachmentHands, (hand => HandChiralityToHandedness(hand.chirality) == ControllerHandedness));
+
+                // Enable all attachment point flags in the leap hand. By default, only the wrist and the palm are enabled.
+                foreach (TrackedHandJoint joint in Enum.GetValues(typeof(TrackedHandJoint)))
+                {
+                    AttachmentHands.attachmentPoints |= ConvertMRTKJointToLeapJoint(joint);
+                }            
+            }
+        }
+#endif
+        public override bool TryGetJoint(TrackedHandJoint joint, out MixedRealityPose pose)
+        {
+#if LEAPMOTIONCORE_PRESENT
+            if (attachmentHand != null && attachmentHand.isTracked)
+            {
+                // Is the current joint a metacarpal
+                bool isMetacarpal = Array.Exists(metacarpals, (metacarpalJoint => metacarpalJoint == joint));
```

Comment:
Be careful with anonymous lambdas like this, which can lead to GC overhead. I know that there have been several articles around this and Unity in the past, so I've tended to avoid this.

If we want to do this, it would be good to do some profiling to make sure that the data suggests this is safe to do.

Consider that TryGetJoint will probably get called in an inner loop update, so it has to be performant.

This could probably be avoided if metacarpals wasn't an array but was a set instead.

Reply:
Removed the anon lambdas